# Evaluación final - Módulo 3 Adalab

Nombre: Zara Valentinova


In [1]:
# Importación de librerías

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.max_columns', None)

# from scipy.stats import shapiro, kstest


In [56]:
flights=pd.read_csv("Customer Flight Activity.csv")
loyalty=pd.read_csv("Customer Loyalty History.csv")

## 1. Exploración inicial

Primero vemos las características de las variables que constituyen cada conjunto de datos.

### 1.1. Exploramos flights y vemos que:

- Nuestro DataFrame contiene 405624 filas y 10 columnas
- Solo contiene variables numéricas.
- No hay valores nulos, pero si muchos ceros.
- Hay 3712 filas duplicadas. No vemos un motivo para mantener los duplicados, así que los borramos, dejando solo la primera aparición de cada. Así, nuestro DataFrame pasa a ser de 403760 filas.
- Vemos que flights contiene valores duplicados en la columna Loyalty Number. Esto es porque el DataFrame incluye registros del 2017 y 2018 por mes, con lo cual hay clientes que han volado con la compañía ambos años o en diferentes meses/constan en el registro de cada uno de los años aun si no han volado ambos.

In [57]:
flights.shape

(405624, 10)

In [58]:
flights.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 405624 entries, 0 to 405623
Data columns (total 10 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Year                         405624 non-null  int64  
 2   Month                        405624 non-null  int64  
 3   Flights Booked               405624 non-null  int64  
 4   Flights with Companions      405624 non-null  int64  
 5   Total Flights                405624 non-null  int64  
 6   Distance                     405624 non-null  int64  
 7   Points Accumulated           405624 non-null  float64
 8   Points Redeemed              405624 non-null  int64  
 9   Dollar Cost Points Redeemed  405624 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 30.9 MB


In [60]:
flights.isna().sum()

Loyalty Number                 0
Year                           0
Month                          0
Flights Booked                 0
Flights with Companions        0
Total Flights                  0
Distance                       0
Points Accumulated             0
Points Redeemed                0
Dollar Cost Points Redeemed    0
dtype: int64

In [59]:
flights.duplicated().any()


np.True_

In [ ]:
flights = flights.drop_duplicates(keep='first')
flights


In [80]:
flights["Loyalty Number"].unique()

array([100018, 100102, 100140, ..., 999731, 999788, 999891],
      shape=(16737,))

In [11]:
flights.describe().T

count           mean            std       min  \
Loyalty Number               405624.0  550037.873084  258935.286969  100018.0   
Year                         405624.0    2017.500000       0.500001    2017.0   
Month                        405624.0       6.500000       3.452057       1.0   
Flights Booked               405624.0       4.115052       5.225518       0.0   
Flights with Companions      405624.0       1.031805       2.076869       0.0   
Total Flights                405624.0       5.146858       6.521227       0.0   
Distance                     405624.0    1208.880059    1433.155320       0.0   
Points Accumulated           405624.0     123.692721     146.599831       0.0   
Points Redeemed              405624.0      30.696872     125.486049       0.0   
Dollar Cost Points Redeemed  405624.0       2.484503      10.150038       0.0   

                                   25%       50%        75%       max  
Loyalty Number               326961.00  550834.0  772194.00  999986.0  
Year                           2017.00    2017.5    2018.00    2018.0  
Month                             3.75       6.5       9.25      12.0  
Flights Booked                    0.00       1.0       8.00      21.0  
Flights with Companions           0.00       0.0       1.00      11.0  
Total Flights                     0.00       1.0      10.00      32.0  
Distance                          0.00     488.0    2336.00    6293.0  
Points Accumulated                0.00      50.0     239.00     676.5  
Points Redeemed                   0.00       0.0       0.00     876.0  
Dollar Cost Points Redeemed       0.00       0.0       0.00      71.0

In [ ]:
flights["Year"]

array([2017, 2018])

In [87]:
flights2017= flights[flights["Year"] == 2017]
flights2017

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
0               100018  2017      1               3                        0   
1               100102  2017      1              10                        4   
2               100140  2017      1               6                        0   
3               100214  2017      1               0                        0   
4               100272  2017      1               0                        0   
...                ...   ...    ...             ...                      ...   
405484          999891  2017     12               0                        0   
405488          356460  2017      6              12                        4   
405530          663150  2017      3              10                        3   
405564          313477  2017      1               8                        3   
405579          205087  2017     11               8                        3   

        Total Flights  Distance  Points Accumulated  Points Redeemed  \
0                   3      1521               152.0                0   
1                  14      2030               203.0                0   
2                   6      1200               120.0                0   
3                   0         0                 0.0                0   
4                   0         0                 0.0                0   
...               ...       ...                 ...              ...   
405484              0         0                 0.0                0   
405488             16      1024               102.0              823   
405530             13       832                83.0              681   
405564             11       704                70.0              548   
405579             11       704                70.0              571   

        Dollar Cost Points Redeemed  
0                                 0  
1                                 0  
2                                 0  
3                                 0  
4                                 0  
...                             ...  
405484                            0  
405488                           66  
405530                           55  
405564                           44  
405579                           46  

[201738 rows x 10 columns]

In [88]:
flights2017[flights2017.duplicated(subset="Loyalty Number", keep=False)]


Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
0               100018  2017      1               3                        0   
1               100102  2017      1              10                        4   
2               100140  2017      1               6                        0   
3               100214  2017      1               0                        0   
4               100272  2017      1               0                        0   
...                ...   ...    ...             ...                      ...   
405484          999891  2017     12               0                        0   
405488          356460  2017      6              12                        4   
405530          663150  2017      3              10                        3   
405564          313477  2017      1               8                        3   
405579          205087  2017     11               8                        3   

        Total Flights  Distance  Points Accumulated  Points Redeemed  \
0                   3      1521               152.0                0   
1                  14      2030               203.0                0   
2                   6      1200               120.0                0   
3                   0         0                 0.0                0   
4                   0         0                 0.0                0   
...               ...       ...                 ...              ...   
405484              0         0                 0.0                0   
405488             16      1024               102.0              823   
405530             13       832                83.0              681   
405564             11       704                70.0              548   
405579             11       704                70.0              571   

        Dollar Cost Points Redeemed  
0                                 0  
1                                 0  
2                                 0  
3                                 0  
4                                 0  
...                             ...  
405484                            0  
405488                           66  
405530                           55  
405564                           44  
405579                           46  

[201738 rows x 10 columns]

In [94]:
# flights2018= flights[flights["Year"] == 2018]
# flights2018

In [90]:
flights2018[flights2018.duplicated(subset="Loyalty Number", keep=False)]

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
27              691626  2018      9               8                        7   
87              627915  2018     11               7                        6   
102             471706  2018      7              12                       10   
132             516133  2018     11              10                        6   
165             409051  2018      9              11                        7   
...                ...   ...    ...             ...                      ...   
405619          999902  2018     12               0                        0   
405620          999911  2018     12               0                        0   
405621          999940  2018     12               3                        0   
405622          999982  2018     12               0                        0   
405623          999986  2018     12               0                        0   

        Total Flights  Distance  Points Accumulated  Points Redeemed  \
27                 15      4245               424.0              341   
87                 13      3757               375.0              310   
102                22      5896               589.0              445   
132                16      3936               393.0              312   
165                18      4428               442.0              343   
...               ...       ...                 ...              ...   
405619              0         0                 0.0                0   
405620              0         0                 0.0                0   
405621              3      1233               123.0                0   
405622              0         0                 0.0                0   
405623              0         0                 0.0                0   

        Dollar Cost Points Redeemed  
27                               28  
87                               25  
102                              36  
132                              25  
165                              28  
...                             ...  
405619                            0  
405620                            0  
405621                            0  
405622                            0  
405623                            0  

[202022 rows x 10 columns]

### 1.2. Exploramos loyalty y vemos que:

- Nuestro DataFrame contiene 16737 filas y 16 columnas
- Contiene siete variables numéricas y nueve de tipo objeto.
- Tres de las variables contienen valores nulos: Salary (4238 nulos), Cancellation Year y Cancellation Month (14670 nulos cada). Esto podría indicar que estos usuarios siguen activos.
- No hay filas duplicadas.
- Al explorar las principales estadísticas, vemos que la columna Salary contiene valores negativos, lo cual tendremos que corregir. 
- También confirmamos que la columna Loyalty Number solo contiene valores únicos (como debería ser dado que es un número identificador).

In [77]:
loyalty.shape

(16737, 16)

In [76]:
loyalty.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16737 entries, 0 to 16736
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Loyalty Number      16737 non-null  int64  
 1   Country             16737 non-null  object 
 2   Province            16737 non-null  object 
 3   City                16737 non-null  object 
 4   Postal Code         16737 non-null  object 
 5   Gender              16737 non-null  object 
 6   Education           16737 non-null  object 
 7   Salary              12499 non-null  float64
 8   Marital Status      16737 non-null  object 
 9   Loyalty Card        16737 non-null  object 
 10  CLV                 16737 non-null  float64
 11  Enrollment Type     16737 non-null  object 
 12  Enrollment Year     16737 non-null  int64  
 13  Enrollment Month    16737 non-null  int64  
 14  Cancellation Year   2067 non-null   float64
 15  Cancellation Month  2067 non-null   float64
dtypes: f

In [75]:
loyalty.isna().sum()

Loyalty Number            0
Country                   0
Province                  0
City                      0
Postal Code               0
Gender                    0
Education                 0
Salary                 4238
Marital Status            0
Loyalty Card              0
CLV                       0
Enrollment Type           0
Enrollment Year           0
Enrollment Month          0
Cancellation Year     14670
Cancellation Month    14670
dtype: int64

In [74]:
loyalty.duplicated().any()

np.False_

In [8]:
loyalty.describe().T

count           mean            std        min  \
Loyalty Number      16737.0  549735.880445  258912.132453  100018.00   
Salary              12499.0   79245.609409   35008.297285  -58486.00   
CLV                 16737.0    7988.896536    6860.982280    1898.01   
Enrollment Year     16737.0    2015.253211       1.979111    2012.00   
Enrollment Month    16737.0       6.669116       3.398958       1.00   
Cancellation Year    2067.0    2016.503145       1.380743    2013.00   
Cancellation Month   2067.0       6.962748       3.455297       1.00   

                          25%        50%        75%        max  
Loyalty Number      326603.00  550434.00  772019.00  999986.00  
Salary               59246.50   73455.00   88517.50  407228.00  
CLV                   3980.84    5780.18    8940.58   83325.38  
Enrollment Year       2014.00    2015.00    2017.00    2018.00  
Enrollment Month         4.00       7.00      10.00      12.00  
Cancellation Year     2016.00    2017.00    2018.00    2018.00  
Cancellation Month       4.00       7.00      10.00      12.00

## 1.3. Unión de los dos conjuntos de datos

Uniremos los dos conjuntos de datos a través de la columna Loyalty History. El número de valores únicos de la columna en flights coincide con la que está en  loyalty, indicando que ninguno de los dos contiene información sobre usuarios que no se encuentran en el otro.